crates/chain/src/tx_data_traits.rs

```diff
@@ -31,3 +33,79 @@ impl ForEachTxOut for Transaction {
         }
     }
 }
+
+/// Trait that "anchors" blockchain data in a specific block of height and hash.
+///
+/// This trait is typically associated with blockchain data such as transactions.
+pub trait BlockAnchor:
+    core::fmt::Debug + Clone + Eq + PartialOrd + Ord + core::hash::Hash + Send + Sync + 'static
+{
+    /// Returns the [`BlockId`] that the associated blockchain data is "anchored" in.
+    fn anchor_block(&self) -> BlockId;
+}
+
+impl<A: BlockAnchor> BlockAnchor for &'static A {
+    fn anchor_block(&self) -> BlockId {
+        <A as BlockAnchor>::anchor_block(self)
+    }
+}
+
+impl BlockAnchor for (u32, BlockHash) {
+    fn anchor_block(&self) -> BlockId {
+        (*self).into()
+    }
+}
+
+/// Represents a service that tracks the best chain history.
+/// TODO: How do we ensure the chain oracle is consistent across a single call?
+/// * We need to somehow lock the data! What if the ChainOracle is remote?
+/// * Get tip method! And check the tip still exists at the end! And every internal call
+///   does not go beyond the initial tip.
+pub trait ChainOracle {
+    /// Error type.
+    type Error: core::fmt::Debug;
+
+    /// Returns the block hash (if any) of the given `height`.
+    fn get_block_in_best_chain(&self, height: u32) -> Result<Option<BlockHash>, Self::Error>;
+
+    /// Determines whether the block of [`BlockId`] exists in the best chain.
+    fn is_block_in_best_chain(&self, block_id: BlockId) -> Result<bool, Self::Error> {
+        Ok(matches!(self.get_block_in_best_chain(block_id.height)?, Some(h) if h == block_id.hash))
+    }
+}
+
+// [TODO] We need stuff for smart pointers. Maybe? How does rust lib do this?
+// Box<dyn ChainOracle>, Arc<dyn ChainOracle> ????? I will figure it out
+impl<C: ChainOracle> ChainOracle for &C {
+    type Error = C::Error;
+
+    fn get_block_in_best_chain(&self, height: u32) -> Result<Option<BlockHash>, Self::Error> {
+        <C as ChainOracle>::get_block_in_best_chain(self, height)
+    }
+
+    fn is_block_in_best_chain(&self, block_id: BlockId) -> Result<bool, Self::Error> {
+        <C as ChainOracle>::is_block_in_best_chain(self, block_id)
+    }
+}
+
+/// Represents an index of transaction data.
+pub trait TxIndex {
```

Comment:
We could name it `SpkIndex` because that's what these methods indexes the transactions with. Or we can clarify that in the doc. It's not readily apparent what this kind of Indexing is going on here, unless the reader looks into the impls of this trait.

Reply:
> script pubkey indexes are only one type of transaction index. This trait should not have is_spk_owned.

That makes sense to me. So is this intended to be a generic `TxIndex`, or `SpkIndex`, which, as you said, is _one type of indexing_. Maybe clarifying it a bit in docs would help?

Agreed that this fits better in the `indexed_tx_graph.rs`.